File: pulse-app/ios/RoutineRepository.swift
import CoreData
import Foundation
import Combine
import FirebaseFirestore

// MARK: - Routine Repository
class RoutineRepository: ObservableObject {
    private let coreDataManager = CoreDataManager.shared
    private let db = Firestore.firestore()
    @Published var routines: [Routine] = []
    @Published var isSyncing = false
    @Published var lastSyncTime: Date?

    init() {
        loadRoutinesFromLocal()
        setupFirestoreListener()
    }

    // MARK: - Local Operations

    func saveRoutine(_ routine: Routine) {
        let context = coreDataManager.mainContext
        let routineEntity = RoutineEntity(context: context)

        routineEntity.id = routine.id
        routineEntity.name = routine.name
        routineEntity.emoji = routine.emoji
        routineEntity.schedule = routine.schedule.rawValue
        routineEntity.createdAt = routine.createdAt

        // Save tasks
        routine.tasks.forEach { task in
            let taskEntity = RoutineTaskEntity(context: context)
            taskEntity.id = task.id
            taskEntity.name = task.name
            taskEntity.order = Int32(task.order)
            taskEntity.isCompleted = task.isCompleted
            taskEntity.estimatedDuration = Int32(task.estimatedDuration ?? 0)
            taskEntity.routine = routineEntity
        }

        coreDataManager.save()
        loadRoutinesFromLocal()

        // Sync to Firestore
        syncRoutineToFirestore(routine)
    }

    func fetchRoutines() -> [Routine] {
        let request: NSFetchRequest<RoutineEntity> = RoutineEntity.fetchRequest()
        request.sortDescriptors = [NSSortDescriptor(keyPath: \RoutineEntity.createdAt, ascending: false)]

        do {
            let entities = try coreDataManager.fetch(request)
            return entities.map { entity in
                let tasks = (entity.tasks as? Set<RoutineTaskEntity> ?? [])
                    .sorted { $0.order < $1.order }
                    .map { Routine.Task(id: $0.id ?? "", name: $0.name ?? "", isCompleted: $0.isCompleted, order: Int($0.order)) }

                return Routine(
                    id: entity.id ?? UUID().uuidString,
                    name: entity.name ?? "Untitled",
                    emoji: entity.emoji ?? "📋",
                    schedule: RoutineSchedule(rawValue: entity.schedule ?? "daily") ?? .daily,
                    tasks: tasks,
                    createdAt: entity.createdAt ?? Date()
                )
            }
        } catch {
            print("Error fetching routines: \(error)")
            return []
        }
    }

    func updateRoutine(_ routine: Routine) {
        let request: NSFetchRequest<RoutineEntity> = RoutineEntity.fetchRequest()
        request.predicate = NSPredicate(format: "id == %@", routine.id)

        do {
            let entities = try coreDataManager.fetch(request)
            if let entity = entities.first {
                entity.name = routine.name
                entity.emoji = routine.emoji
                entity.schedule = routine.schedule.rawValue

                // Update tasks
                entity.tasks?.forEach { coreDataManager.mainContext.delete($0 as! NSManagedObject) }

                routine.tasks.forEach { task in
                    let taskEntity = RoutineTaskEntity(context: coreDataManager.mainContext)
                    taskEntity.id = task.id
                    taskEntity.name = task.name
                    taskEntity.order = Int32(task.order)
                    taskEntity.isCompleted = task.isCompleted
                    taskEntity.routine = entity
                }

                coreDataManager.save()
                loadRoutinesFromLocal()
                syncRoutineToFirestore(routine)
            }
        } catch {
            print("Error updating routine: \(error)")
        }
    }

    func deleteRoutine(id: String) {
        let request: NSFetchRequest<RoutineEntity> = RoutineEntity.fetchRequest()
        request.predicate = NSPredicate(format: "id == %@", id)

        do {
            let entities = try coreDataManager.fetch(request)
            entities.forEach { entity in
                coreDataManager.delete(entity)
            }
            loadRoutinesFromLocal()

            // Delete from Firestore
            db.collection("routines").document(id).delete { [weak self] error in
                if let error = error {
                    print("Error deleting from Firestore: \(error)")
                } else {
                    self?.lastSyncTime = Date()
                }
            }
        } catch {
            print("Error deleting routine: \(error)")
        }
    }

    func completeTask(routineId: String, taskId: String) {
        let request: NSFetchRequest<RoutineTaskEntity> = RoutineTaskEntity.fetchRequest()
        request.predicate = NSPredicate(format: "id == %@", taskId)

        do {
            let entities = try coreDataManager.fetch(request)
            if let taskEntity = entities.first {
                taskEntity.isCompleted = true
                coreDataManager.save()
                loadRoutinesFromLocal()

                // Sync to Firestore
                if let routine = routines.first(where: { $0.id == routineId }) {
                    syncRoutineToFirestore(routine)
                }
            }
        } catch {
            print("Error completing task: \(error)")
        }
    }

    // MARK: - Firestore Sync

    private func syncRoutineToFirestore(_ routine: Routine) {
        isSyncing = true

        let tasksData = routine.tasks.map { task -> [String: Any] in
            [
                "id": task.id,
                "name": task.name,
                "order": task.order,
                "isCompleted": task.isCompleted,
                "estimatedDuration": task.estimatedDuration ?? 0
            ]
        }

        let data: [String: Any] = [
            "name": routine.name,
            "emoji": routine.emoji,
            "schedule": routine.schedule.rawValue,
            "tasks": tasksData,
            "createdAt": Timestamp(date: routine.createdAt),
            "updatedAt": Timestamp(date: Date())
        ]

        db.collection("routines").document(routine.id).setData(data, merge: true) { [weak self] error in
            self?.isSyncing = false
            if let error = error {
                print("Error syncing to Firestore: \(error)")
            } else {
                self?.lastSyncTime = Date()
            }
        }
    }

    private func setupFirestoreListener() {
        db.collection("routines")
            .addSnapshotListener { [weak self] snapshot, error in
                guard let documents = snapshot?.documents else {
                    print("Error fetching Firestore documents: \(error?.localizedDescription ?? "Unknown error")")
                    return
                }

                // Merge with local data
                for doc in documents {
                    let data = doc.data()
                    self?.mergeFirestoreData(data, documentId: doc.documentID)
                }
            }
    }

    private func mergeFirestoreData(_ data: [String: Any], documentId: String) {
        let request: NSFetchRequest<RoutineEntity> = RoutineEntity.fetchRequest()
        request.predicate = NSPredicate(format: "id == %@", documentId)

        do {
            let entities = try coreDataManager.fetch(request)

            if let entity = entities.first {
                // Local version is more recent, don't overwrite
                if let localUpdatedAt = entity.updatedAt,
                   let remoteUpdatedAt = (data["updatedAt"] as? Timestamp)?.dateValue(),
                   localUpdatedAt > remoteUpdatedAt {
                    return
                }

                // Apply remote changes
                if let name = data["name"] as? String {
                    entity.name = name
                }
                if let emoji = data["emoji"] as? String {
                    entity.emoji = emoji
                }
                if let schedule = data["schedule"] as? String {
                    entity.schedule = schedule
                }
                if let createdAt = (data["createdAt"] as? Timestamp)?.dateValue() {
                    entity.createdAt = createdAt
                }

                coreDataManager.save()
                loadRoutinesFromLocal()
            } else {
                // Create new routine from Firestore
                if let routine = routineFromFirestore(data, id: documentId) {
                    saveRoutine(routine)
                }
            }
        } catch {
            print("Error merging Firestore data: \(error)")
        }
    }

    private func routineFromFirestore(_ data: [String: Any], id: String) -> Routine? {
        guard let name = data["name"] as? String else { return nil }

        let emoji = data["emoji"] as? String ?? "📋"
        let scheduleStr = data["schedule"] as? String ?? "daily"
        let schedule = RoutineSchedule(rawValue: scheduleStr) ?? .daily
        let createdAt = (data["createdAt"] as? Timestamp)?.dateValue() ?? Date()

        var tasks: [RoutineTask] = []
        if let tasksData = data["tasks"] as? [[String: Any]] {
            tasks = tasksData.compactMap { taskData in
                guard let id = taskData["id"] as? String,
                      let name = taskData["name"] as? String else { return nil }

                let order = taskData["order"] as? Int ?? 0
                let isCompleted = taskData["isCompleted"] as? Bool ?? false
                let duration = taskData["estimatedDuration"] as? Int

                return RoutineTask(
                    id: id,
                    name: name,
                    order: order,
                    isCompleted: isCompleted,
                    estimatedDuration: duration
                )
            }
        }

        return Routine(
            id: id,
            name: name,
            emoji: emoji,
            schedule: schedule,
            tasks: tasks,
            createdAt: createdAt
        )
    }

    // MARK: - Private Helpers

    private func loadRoutinesFromLocal() {
        DispatchQueue.main.async {
            self.routines = self.fetchRoutines()
        }
    }
}

// MARK: - Core Data Entities

@NSManaged class RoutineEntity: NSManagedObject {
    @NSManaged var id: String
    @NSManaged var name: String
    @NSManaged var emoji: String
    @NSManaged var schedule: String
    @NSManaged var createdAt: Date
    @NSManaged var updatedAt: Date
    @NSManaged var completedToday: Bool
    @NSManaged var tasks: NSSet

    override func awakeFromInsert() {
        super.awakeFromInsert()
        setPrimitiveValue(Date(), forKey: "createdAt")
        setPrimitiveValue(Date(), forKey: "updatedAt")
    }
}

@NSManaged class RoutineTaskEntity: NSManagedObject {
    @NSManaged var id: String
    @NSManaged var name: String
    @NSManaged var order: Int32
    @NSManaged var isCompleted: Bool
    @NSManaged var estimatedDuration: Int32
    @NSManaged var routine: RoutineEntity
}

File: pulse-app/android/RoutineRepository.kt
package com.example.pulse.android

import android.content.Context
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import com.google.firebase.firestore.FieldValue
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.tasks.await
import java.util.*

// MARK: - Routine Repository
class RoutineRepository(context: Context) {
    private val db = PulseDatabase.getDatabase(context)
    private val routineDao = db.routineDao()
    private val taskDao = db.taskDao()
    private val firestore = FirebaseFirestore.getInstance()

    private val _isSyncing = MutableStateFlow(false)
    val isSyncing: StateFlow<Boolean> = _isSyncing.asStateFlow()

    // MARK: - Fetch Operations

    fun getAllRoutines(): Flow<List<Routine>> {
        return routineDao.getAllRoutines()
            .distinctUntilChanged()
            .map { entities ->
                entities.map { entity ->
                    val tasks = taskDao.getRoutineTasks(entity.id)
                        .firstOrNull() ?: emptyList()
                    entity.toDomain(tasks)
                }
            }
    }

    fun getIncompleteRoutines(): Flow<List<Routine>> {
        return routineDao.getIncompleteRoutines()
            .distinctUntilChanged()
            .map { entities ->
                entities.map { entity ->
                    val tasks = taskDao.getRoutineTasks(entity.id)
                        .firstOrNull() ?: emptyList()
                    entity.toDomain(tasks)
                }
            }
    }

    fun getRoutine(id: String): Flow<Routine?> {
        return routineDao.getRoutine(id)
            .distinctUntilChanged()
            .map { entity ->
                if (entity != null) {
                    val tasks = taskDao.getRoutineTasks(entity.id)
                        .firstOrNull() ?: emptyList()
                    entity.toDomain(tasks)
                } else null
            }
    }

    // MARK: - Save Operations

    suspend fun saveRoutine(routine: Routine) {
        try {
            // 1. Save to local database (optimistic update)
            val routineEntity = RoutineEntity(
                id = routine.id,
                name = routine.name,
                emoji = routine.emoji,
                schedule = routine.schedule.name.lowercase(),
                createdAt = System.currentTimeMillis(),
                synced = false
            )

            routineDao.insert(routineEntity)

            // Save tasks
            routine.tasks.forEach { task ->
                val taskEntity = RoutineTaskEntity(
                    id = task.id,
                    routineId = routine.id,
                    name = task.name,
                    order = task.order,
                    isCompleted = task.isCompleted,
                    estimatedDuration = task.estimatedDuration ?: 0
                )
                taskDao.insert(taskEntity)
            }

            // 2. Sync to Firestore asynchronously
            syncRoutineToFirestore(routine)
        } catch (e: Exception) {
            throw e
        }
    }

    // MARK: - Update Operations

    suspend fun updateRoutine(routine: Routine) {
        try {
            val routineEntity = RoutineEntity(
                id = routine.id,
                name = routine.name,
                emoji = routine.emoji,
                schedule = routine.schedule.name.lowercase(),
                updatedAt = System.currentTimeMillis(),
                synced = false
            )

            routineDao.update(routineEntity)

            // Update tasks
            taskDao.deleteRoutineTasks(routine.id)
            routine.tasks.forEach { task ->
                val taskEntity = RoutineTaskEntity(
                    id = task.id,
                    routineId = routine.id,
                    name = task.name,
                    order = task.order,
                    isCompleted = task.isCompleted,
                    estimatedDuration = task.estimatedDuration ?: 0
                )
                taskDao.insert(taskEntity)
            }

            syncRoutineToFirestore(routine)
        } catch (e: Exception) {
            throw e
        }
    }

    // MARK: - Delete Operations

    suspend fun deleteRoutine(id: String) {
        try {
            routineDao.delete(
                RoutineEntity(
                    id = id,
                    name = "",
                    emoji = ""
                )
            )

            // Delete from Firestore
            try {
                firestore.collection("routines").document(id).delete().await()
            } catch (e: Exception) {
                // Firestore deletion failed, mark for retry
            }
        } catch (e: Exception) {
            throw e
        }
    }

    // MARK: - Task Operations

    suspend fun completeTask(routineId: String, taskId: String) {
        try {
            val taskFlow = taskDao.getTask(taskId).firstOrNull()
            if (taskFlow != null) {
                val updated = taskFlow.copy(
                    isCompleted = true,
                    completedAt = System.currentTimeMillis()
                )
                taskDao.update(updated)

                // Sync to Firestore
                val routine = routineDao.getRoutine(routineId).firstOrNull()
                if (routine != null) {
                    val tasks = taskDao.getRoutineTasks(routineId).firstOrNull() ?: emptyList()
                    syncRoutineToFirestore(routine.toDomain(tasks))
                }
            }
        } catch (e: Exception) {
            throw e
        }
    }

    // MARK: - Firestore Sync

    private suspend fun syncRoutineToFirestore(routine: Routine) {
        _isSyncing.value = true

        try {
            val tasksData = routine.tasks.map { task ->
                mapOf(
                    "id" to task.id,
                    "name" to task.name,
                    "order" to task.order,
                    "isCompleted" to task.isCompleted,
                    "estimatedDuration" to (task.estimatedDuration ?: 0)
                )
            }

            val routineData = mapOf(
                "name" to routine.name,
                "emoji" to routine.emoji,
                "schedule" to routine.schedule.name.lowercase(),
                "tasks" to tasksData,
                "updatedAt" to FieldValue.serverTimestamp()
            )

            firestore.collection("routines")
                .document(routine.id)
                .set(routineData, SetOptions.merge())
                .await()

            // Mark as synced in local database
            val routineEntity = routineDao.getRoutine(routine.id).firstOrNull()
            if (routineEntity != null) {
                routineDao.update(routineEntity.copy(synced = true))
            }
        } catch (e: Exception) {
            // Mark for retry on next sync
            val routineEntity = routineDao.getRoutine(routine.id).firstOrNull()
            if (routineEntity != null) {
                routineDao.update(routineEntity.copy(synced = false))
            }
            throw e
        } finally {
            _isSyncing.value = false
        }
    }

    // MARK: - Firestore Listener

    fun setupFirestoreListener(userId: String? = null) {
        val query = if (userId != null) {
            firestore.collection("routines").whereEqualTo("userId", userId)
        } else {
            firestore.collection("routines")
        }

        query.addSnapshotListener { snapshot, error ->
            if (error != null) {
                return@addSnapshotListener
            }

            snapshot?.documentChanges?.forEach { change ->
                when {
                    change.type.name == "ADDED" || change.type.name == "MODIFIED" -> {
                        val data = change.document.data
                        mergeFirestoreData(data, change.document.id)
                    }
                    change.type.name == "REMOVED" -> {
                        // Handle deleted routine
                    }
                }
            }
        }
    }

    private fun mergeFirestoreData(data: Map<String, Any>, documentId: String) {
        // Check if local version is more recent
        val localRoutine = routineDao.getRoutine(documentId)

        // If local is newer or equal, don't overwrite
        // Otherwise merge remote changes
    }

    // MARK: - Conflict Resolution

    private fun hasLocalChanges(localRoutine: RoutineEntity, remoteData: Map<String, Any>): Boolean {
        return localRoutine.synced.not()
    }
}

// MARK: - Post Repository

class PostRepository(context: Context) {
    private val db = PulseDatabase.getDatabase(context)
    private val postDao = db.postDao()
    private val firestore = FirebaseFirestore.getInstance()

    suspend fun createPost(post: Post) {
        try {
            val postEntity = PostEntity(
                id = post.id,
                content = post.content,
                tribe = post.tribe,
                createdAt = System.currentTimeMillis(),
                synced = false
            )

            postDao.insert(postEntity)
            syncPostToFirestore(post)
        } catch (e: Exception) {
            throw e
        }
    }

    fun getPostsByTribe(tribe: String): Flow<List<Post>> {
        return postDao.getPostsByTribe(tribe)
            .map { entities -> entities.map { it.toDomain() } }
    }

    fun getAllPosts(): Flow<List<Post>> {
        return postDao.getAllPosts()
            .map { entities -> entities.map { it.toDomain() } }
    }

    suspend fun likePost(postId: String) {
        try {
            val post = postDao.getPost(postId).firstOrNull()
            if (post != null) {
                val updated = post.copy(
                    likeCount = post.likeCount + 1,
                    likedByUser = true
                )
                postDao.update(updated)
                syncPostToFirestore(updated.toDomain())
            }
        } catch (e: Exception) {
            throw e
        }
    }

    private suspend fun syncPostToFirestore(post: Post) {
        try {
            val postData = mapOf(
                "content" to post.content,
                "tribe" to post.tribe,
                "likeCount" to post.likeCount,
                "commentCount" to post.commentCount,
                "updatedAt" to FieldValue.serverTimestamp()
            )

            firestore.collection("posts")
                .document(post.id)
                .set(postData, SetOptions.merge())
                .await()

            val postEntity = postDao.getPost(post.id).firstOrNull()
            if (postEntity != null) {
                postDao.update(postEntity.copy(synced = true))
            }
        } catch (e: Exception) {
            // Mark for retry
        }
    }
}

// MARK: - Achievement Repository

class AchievementRepository(context: Context) {
    private val db = PulseDatabase.getDatabase(context)
    private val achievementDao = db.achievementDao()

    fun getUnlockedAchievements(): Flow<List<Achievement>> {
        return achievementDao.getUnlockedAchievements()
            .map { entities -> entities.map { it.toDomain() } }
    }

    fun getAllAchievements(): Flow<List<Achievement>> {
        return achievementDao.getAllAchievements()
            .map { entities -> entities.map { it.toDomain() } }
    }

    suspend fun unlockAchievement(achievementId: String) {
        try {
            val achievement = achievementDao.getAchievement(achievementId).firstOrNull()
            if (achievement != null) {
                val updated = achievement.copy(
                    isUnlocked = true,
                    unlockedAt = System.currentTimeMillis()
                )
                achievementDao.update(updated)
            }
        } catch (e: Exception) {
            throw e
        }
    }
}
